### PR TITLE
added temporal units to support binning by decade, hour, and minute

### PR DESCRIPTION
**Pull request by dcy2003**
#555

### CODE REVIEW

**Comment by rfecher** [comment]
To expose it to the commandline tools you have t o add it as options to the periodicity in SpatialTemporalDimensionalityTypeProvider.  Also, if you could figure out a way to get "week" in there that would be awesome!

File: core/geotime/src/main/java/mil/nga/giat/geowave/core/geotime/index/dimension/TemporalBinningStrategy.java [hunk 1]
```diff
@@ -211,6 +237,17 @@ public int getFixedBinIdSize() {
...
+				return StringUtils.stringToBinary((Integer.toString(value.get(Calendar.YEAR)) + "_"
+						+ TWO_DIGIT_NUMBER.format(value.get(Calendar.MONTH)) + "_"
+						+ TWO_DIGIT_NUMBER.format(value.get(Calendar.DAY_OF_MONTH)) + "_" + TWO_DIGIT_NUMBER
+						.format(value.get(Calendar.HOUR_OF_DAY))));
+			case MINUTE:
+				return StringUtils.stringToBinary((Integer.toString(value.get(Calendar.YEAR)) + "_"
+						+ TWO_DIGIT_NUMBER.format(value.get(Calendar.MONTH)) + "_"
+						+ TWO_DIGIT_NUMBER.format(value.get(Calendar.DAY_OF_MONTH)) + "_"
+						+ TWO_DIGIT_NUMBER.format(value.get(Calendar.HOUR_OF_DAY)) + "_" + TWO_DIGIT_NUMBER
+						.format(value.get(Calendar.MINUTE))));
```
Comment (rwgdrummer) [inline core/geotime/src/main/java/mil/nga/giat/geowave/core/geotime/index/dimension/TemporalBinningStrategy.java]: May be we should now consider compressing these a bit.

Comment (rfecher) [inline core/geotime/src/main/java/mil/nga/giat/geowave/core/geotime/index/dimension/TemporalBinningStrategy.java]: perhaps a good idea - a req't for that compression is that for each periodicity/bin, the binary size must be fixed, but I don't think that should be a problem. preserving sort order actually shouldn't matter as we are just using the periodicity/bin as an enumeration

File: core/geotime/src/main/java/mil/nga/giat/geowave/core/geotime/index/dimension/TemporalBinningStrategy.java [hunk 2]
```diff
@@ -223,6 +260,20 @@ private Calendar getStartEpoch(
...
+				cal.set(
+						Calendar.MINUTE,
+						minute);
+			case HOUR:
+				final int hour = Integer.parseInt(str.substring(
+						11,
+						13));
+				cal.set(
+						Calendar.HOUR_OF_DAY,
+						hour);
```
Comment (rwgdrummer) [inline core/geotime/src/main/java/mil/nga/giat/geowave/core/geotime/index/dimension/TemporalBinningStrategy.java]: Should we add DECADE?
Basically, it is  a variant to setting the YEAR.

Comment (rfecher) [inline core/geotime/src/main/java/mil/nga/giat/geowave/core/geotime/index/dimension/TemporalBinningStrategy.java]: it seems that decade is essentially accounted for entirely within getBinSizeMillis() but I kinda agree with @rwgdrummer - a check for decade here would essentially ensure that the year is a multiple of 10 (rounding down if it is now).  Perhaps redundant but it is definitely consistent with what one would expect from getStartEpoch()

Comment (dcy2003) [inline core/geotime/src/main/java/mil/nga/giat/geowave/core/geotime/index/dimension/TemporalBinningStrategy.java]: Good call.  Will do.

File: core/geotime/src/main/java/mil/nga/giat/geowave/core/geotime/index/dimension/TemporalBinningStrategy.java (hunk 1 above)
Comment (dcy2003) [inline core/geotime/src/main/java/mil/nga/giat/geowave/core/geotime/index/dimension/TemporalBinningStrategy.java]: So what do we want to do here?  I can come up with an approach to compress these a bit, which will require me to rework the parsing logic within getStartEpoch(), and and adjust the fixed binId sizes.  Are we ok with this being a breaking change, or do we want to add complicated logic to handle both the new and old formats (a bit ugly)?

Comment (rfecher) [inline core/geotime/src/main/java/mil/nga/giat/geowave/core/geotime/index/dimension/TemporalBinningStrategy.java]: I'd suggest you document an approach in an issue but don't worry about implementing it yet - tag the issue as "breaking changes"

**Comment by dcy2003** [comment]
@rfecher not sure I follow your comment about exposing it in SpatialTemporalDimensionalityTypeProvider.  Isn't that handled via:
https://github.com/dcy2003/geowave/blob/GEOWAVE-555/core/geotime/src/main/java/mil/nga/giat/geowave/core/geotime/ingest/SpatialTemporalDimensionalityTypeProvider.java#L136-138
and parsed via:
https://github.com/dcy2003/geowave/blob/GEOWAVE-555/core/geotime/src/main/java/mil/nga/giat/geowave/core/geotime/ingest/SpatialTemporalDimensionalityTypeProvider.java#L214-232

File: core/geotime/src/main/java/mil/nga/giat/geowave/core/geotime/index/dimension/TemporalBinningStrategy.java (hunk 1 above)
Comment (dcy2003) [inline core/geotime/src/main/java/mil/nga/giat/geowave/core/geotime/index/dimension/TemporalBinningStrategy.java]: Opened #802

**Pull request merged**